File: android/app/src/main/java/com/lilru/liftr/workout/StrengthFinishIncompleteCounts.kt
package com.lilru.liftr.workout

import com.lilru.liftr.ui.active.ActiveStrengthExerciseLine
import java.util.Locale

data class StrengthFinishIncompleteCounts(
    val exercises: Int,
    val sets: Int
) {
    companion object {
        val zero = StrengthFinishIncompleteCounts(0, 0)

        fun aggregate(parts: List<StrengthFinishIncompleteCounts>): StrengthFinishIncompleteCounts {
            return StrengthFinishIncompleteCounts(
                exercises = parts.sumOf { it.exercises },
                sets = parts.sumOf { it.sets }
            )
        }
    }
}

object StrengthFinishIncompleteCounting {
    fun counts(
        exercises: List<ActiveStrengthExerciseLine>,
        completedIndexByExerciseId: Map<Int, Int>
    ): StrengthFinishIncompleteCounts {
        var incompleteSets = 0
        var incompleteExercises = 0
        for (ex in exercises) {
            val total = ex.sets.size
            if (total <= 0) continue
            val done = (completedIndexByExerciseId[ex.workoutExerciseId] ?: 0).coerceIn(0, total)
            val remaining = total - done
            incompleteSets += remaining
            if (done == 0) incompleteExercises += 1
        }
        return StrengthFinishIncompleteCounts(incompleteExercises, incompleteSets)
    }
}

object StrengthFinishConfirmationCopy {
    fun incompleteWarning(exercises: Int, sets: Int): String {
        val lang = Locale.getDefault().language
        return if (lang == "es") {
            "Tienes $exercises ejercicios y $sets series sin terminar. Si finalizas el entrenamiento ahora, se eliminarán automáticamente de tu historial."
        } else {
            "You have $exercises exercises and $sets sets left unfinished. If you finish now, they will be removed from your history automatically."
        }
    }

    fun standardEarlyFinishBody(): String {
        val lang = Locale.getDefault().language
        return if (lang == "es") {
            "No has completado todas las series planificadas. El entrenamiento se guardará solo con las series que hayas hecho. Si finalizas ahora, se publicará automáticamente."
        } else {
            "You haven't completed all planned sets. The workout will be saved with only the sets you actually performed. If you finish now, it will be published automatically."
        }
    }

    fun message(
        incomplete: StrengthFinishIncompleteCounts,
        standardBody: String
    ): String {
        val parts = mutableListOf<String>()
        if (incomplete.exercises > 0 && incomplete.sets > 0) {
            parts += incompleteWarning(incomplete.exercises, incomplete.sets)
        }
        parts += standardBody
        return parts.joinToString("\n\n")
    }
}

File: Liftr/StrengthFinishIncompleteCounts.swift
import Foundation

struct StrengthFinishIncompleteCounts: Equatable {
    let exercises: Int
    let sets: Int

    static let zero = StrengthFinishIncompleteCounts(exercises: 0, sets: 0)

    static func aggregate(_ parts: [StrengthFinishIncompleteCounts]) -> StrengthFinishIncompleteCounts {
        StrengthFinishIncompleteCounts(
            exercises: parts.reduce(0) { $0 + $1.exercises },
            sets: parts.reduce(0) { $0 + $1.sets }
        )
    }
}

enum StrengthFinishIncompleteCounting {
    static func counts<E>(
        exercises: [E],
        totalSets: (E) -> Int,
        completedSetIndex: (E) -> Int
    ) -> StrengthFinishIncompleteCounts {
        var incompleteSets = 0
        var incompleteExercises = 0
        for ex in exercises {
            let total = totalSets(ex)
            guard total > 0 else { continue }
            let done = min(max(0, completedSetIndex(ex)), total)
            let remaining = total - done
            incompleteSets += remaining
            if done == 0 {
                incompleteExercises += 1
            }
        }
        return StrengthFinishIncompleteCounts(exercises: incompleteExercises, sets: incompleteSets)
    }
}

enum StrengthFinishConfirmationCopy {
    static func incompleteWarning(exercises: Int, sets: Int) -> String {
        let lang = Locale.current.language.languageCode?.identifier ?? ""
        if lang == "es" {
            return "Tienes \(exercises) ejercicios y \(sets) series sin terminar. Si finalizas el entrenamiento ahora, se eliminarán automáticamente de tu historial."
        }
        return "You have \(exercises) exercises and \(sets) sets left unfinished. If you finish now, they will be removed from your history automatically."
    }

    static var standardEarlyFinishBody: String {
        let lang = Locale.current.language.languageCode?.identifier ?? ""
        if lang == "es" {
            return "No has completado todas las series planificadas. El entrenamiento se guardará solo con las series que hayas hecho. Si finalizas ahora, se publicará automáticamente."
        }
        return "You haven't completed all planned sets. The workout will be saved with only the sets you actually performed. If you finish now, it will be published automatically."
    }

    static func message(
        incomplete: StrengthFinishIncompleteCounts,
        standardBody: String,
        dualPartnerNote: String? = nil
    ) -> String {
        var parts: [String] = []
        if incomplete.exercises > 0, incomplete.sets > 0 {
            parts.append(incompleteWarning(exercises: incomplete.exercises, sets: incomplete.sets))
        }
        parts.append(standardBody)
        if let dualPartnerNote, !dualPartnerNote.isEmpty {
            parts.append(dualPartnerNote)
        }
        return parts.joined(separator: "\n\n")
    }
}
